### DISCUSSION TITLE
bwc not starting if there is a problem with wifi

### DISCUSSION THREAD

**Discussion by davidmardanielsson** [visualapproach/WiFi-remote-for-Bestway-Lay-Z-SPA#163] [General]
Hi. 

First, thanks for this brilliant solution. Works really well. 

I have the Lazy spa Helsinki and I'm running it 24/7 the whole year. It is located in the countryside of Iceland (actually near a glacier) and I've had issues with the electricity going out a few times. The house is a couple of hours away and I've needed to rely on my neighbour to start the pump again.  Therefore this solution works quite well in that case since it offers to turn on the pump with the same configuration as before. 

I was testing a new wifi connection and I noticed that if I turn on the pump and the wifi doesn't work correctly, the pump itself doesn't turn on. In setup(), it starts with starting the wifi and ends with starting the bwc. 
For me I would like to eliminate the need to actually depend on the wifi, but if the wifi is available it would be used. 
So I was thinking if there would be anything wrong with doing the bwc.begin() before starting the wifi? 

Thanks!

David

**Comment by visualapproach** [comment]
Hi David! Sounds nice to live by a glacier with a hot tub! Maybe post something here ;-) : #13 
I understand your need for wifi independence. The reason wifi.begin() is before bwc.begin() is that the command queue, and therefore the whole program, is depending on the NTP to get the correct time. In principle it's not a problem. In practice there is some more changes needed to make it work. For instance, you can't have the autoportal since it's blocking the program from running. There was a version out there that didn't block, but I didn't get it to work.
I'll look into it and will report here if/when it's done.
/Thomas

**Reply by davidmardanielsson** [reply]
Nice, I will definitely post something to #13. It has been on my list to build some structure around the tub, mainly to insulate it and those photos are definitely an inspiration. 

Ok, so I was going through the code. With the goal of making sure that the pump is started before attempting to initialize the wifi I was thinking if the following makes sense. In setup(), start with setting the clock to some value (a date in the past). Run bwc.begin() where the startNTP() has been split out into an initialize() function which will be run later. Then run bwc.loop one time. 

So:
  ```
time_t rtc = 1510592825;
  timeval tv = { rtc, 0 };
  settimeofday(&tv, nullptr);
  bwc.begin(); // Does not contain the _startNTP()
  bwc.loop(); 
  startWiFi();
  startOTA();
  startServer();
  startWebSocket();
  bwc.initialize(); // Contains the _startNTP()
```

Will this start up the pump?

After this in setup() the startWIFI() and the rest can be run and end with bwc.initialize(); which is basically the NTP and related. 

As mentioned, my main concern is after some power failure the wifi might be in a bad state and for me it is most important that the spa is started up with the default configuration before doing anything else. 

Maybe another option would be to create another function which would send the power / heater / pump commands explicitly?  

I just want to find a clean way to do this because I don't have the means to debug this. 

Thanks!

David.

**Comment by visualapproach** [comment]
I can't tell without testing. I'm working on a fix and will be reporting back

**Reply by davidmardanielsson** [reply]
Amazing, thanks!

**Comment by visualapproach** [comment]
@davidmardanielsson Hey, I updated the development version. I don't know if you can push updates to the glacier from home, but I suggest you test this version before trusting it 100%. I'm running it on my hot tub now and it seems to work as planned, but you'll never know. I tested to turn off wifi for a while and it worked. When I turned on wifi it reconnected fine. See "changelog.txt" file for more info.
Please report how it works for you. If OK I will pull this version to master.

**Reply by davidmardanielsson** [reply]
Hi. Thanks! I just uploaded (I'm actually here now) and I tested the following
* Reboot of the ESP with the wifi working. That works. 
* Power cycle of the spa and wifi off. That does not work still. The screen is off and I waited for a few minutes and nothing happened, even though I enabled the wifi. So I had to enable the wifi and then power cycle the hot tub. 

So my concern is like before, if there is an issue like a power failure and for some reason the wifi doesn't come up properly (even, if the wifi comes up but there is a problem with the internet) the spa needs to start. But also, if there is a power failure and the wifi takes some time to initialize and maybe isn't available at the right time when the spa requires it, the esp goes into a deadlock and isn't able to recover. At least, I tested that for a few minutes and it didn't. Maybe I didn't wait for long enough. 

In the changelog: " The pump will function during the autoportal, but the display will not update", can you elaborate? Would you then be able to push the buttons? Because for me the requirement would be that the spa returns to a normal running state, which is heater on and pump on. 

Thanks for your effort in this. I really appreciate it. 

David.

**Reply by visualapproach** [reply]
This is how it *should* behave:
Wifi off. Power on. The device can't connect so it starts the autoportal which is blocking the main loop from running. However the communication between the pump (cio) and the ESP is interrupt-driven, so it should still work in the sense that the cio thinks there is a display there and it will not throw errors. Display and touch buttons will not work during this time. The autoportal is active for 240 seconds (4 minutes). When it shuts down, the main loop continues, and if you have set the "return to last states" it will do that after another 30 seconds or so.
Every 30 seconds it will try to reconnect to wifi. This will ofc only succeed if the credential has been entered previously at some time.

Your test should be something like this:
Wifi ON. Start device. Enter autoportal and save credentials. From the web GUI Set the pump to ON. Then press ESP RESTART (to force saving of some states which otherwise only saves every couple of hours). Wait for it to restart.
Shut off power and wifi.
Start power and wait at least 5-6 minutes.
The pump should start and the display should wake up and be responsive.

**Reply by davidmardanielsson** [reply]
Ok. 

I did the following: 
* Turned off the spa and wifi, started the spa without wifi turned on and waited for 10 minutes. Nothing happened. Tried this twice. 
* Turned on the wifi after 10 minutes had passed and waited for a few minutes and nothing happened. 
* Power cycled the spa after turning on the wifi and the spa turned on as expected. 

I could plug in a serial console, but can only do it in a few hours. 

Thanks!

**Reply by visualapproach** [reply]
Ok I'll have a look

**Reply by visualapproach** [reply]
Not sure I can do it before you leave. I can confirm that wifi isn't connecting after powerup without wifi, when wifi comes on.

**Reply by davidmardanielsson** [reply]
Bad news, so now the DSP is unresponsive and the web UI also. I can push the unlock key on the DSP, it beeps one time, but nothing else happens. The pump is running, but it seems the heater turned off prematurely.

**Reply by visualapproach** [reply]
Ok I suggest you install the master branch again and I will investigate further. Sorry it didn't work

**Reply by davidmardanielsson** [reply]
But suddenly I can connect to it now and the web UI is working. But the DSP is still off. 

Ok. I'll do that. I would probably want to setup a remote serial console, using a 8266 so I can see what is going. I'll do that next weekend. 

Is your setup working with the dev branch?

Thanks!

**Reply by visualapproach** [reply]
Nope, It's a b*tch to get it to connect after a first failed try. Hope to get it working until your next trip...

**Reply by visualapproach** [reply]
I found out that the autoportal was stopping the device from reconnecting if it didn't find wifi at start-up.  I think it's sorted out now. I've tested to restart the pump with no wifi, and it resumes last states and when I turned on the wifi again, it reconnected fine. (Actually I blacklisted the device in my router temporarily but I hope it's a valid test)

New test procedure for you would be:
Update with dev-version.
Let the device run until the display shows the current temperature without blinking.
Change unit, or pump state or heater state. Then change it to whatever setting you would like it to have. (ESP saves the setting upon something changing)
disconnect wifi and restart pump.
It should now be responsive and restore the last states.
turn on wifi and verify that it connects again.

**Reply by davidmardanielsson** [reply]
Thanks! I'll test it next weekend.

**Comment by davidmardanielsson** [comment]
Hi Thomas. 

One question. The power went out and kind of as expected the pump wouldn't turn on. I suspect that it was my wifi that was too slow to turn on. But I was able to log in to my router and see that there was an IP there that was related to the ESP. So it actually got a DHCP lease and the device name on my router for this IP is ESP-1BB310. I had seen it before but I always used the other IP address that actually did have an open port 80 which was named BW_2.0. But this IP (ESP-1BB310) doesn't have any port open (at least not 80 or 81). 

So I'm wondering why the ESP gets two IP addresses and two leases and what this other IP is used for. And also, can I somehow use this IP to access the ESP?

**Reply by visualapproach** [reply]
Hi!
If running the present development version the hostname should be "SPA-1BB310", but at least my router doesn't care about that. It keeps whatever name it had when it first connected. Which was "ESP-xxxxxxx". Never has there been a hostname "BW_2.0" that I can remember. The MQTT client ID is "BW_2.0.0" and using port 1883 as default. Are both clients really active on your router?
I'm not sure what is going on. Try wiping the flash and reinstall code and data. But before that - go to your device's IP/states.txt and see if it has saved the current states. Should look something like this: {"UNT":1,"HTR":1,"FLT":1}

**Reply by davidmardanielsson** [reply]
Yeah, I'm running master and there might be some stuff from the initial test there that could be interfering. 
I will check this out in a few hours. 

But from the beginning I've been seeing two DHCP leases from the device, and two IP's. I've just been using the one that worked I assumed initially (before I tried OTA) that this was for the OTA upload. 

So I'll do a clean reinstall and see what happens then. I'll also do a test on power outage. I think my issue is actually that when the power goes out completely along with the wifi, my router is a bit slow in making the wifi available. I can imagine that might be the reason for this hanging. 

So to test this on your end, you might want to take the power of your wifi and the spa at the same time and see what happens when the power is back on. I would have loved it this would have worked. It feels so weird because I can actually see this second IP on my router and it actually gets a lease back if I drop it manually. Which means the device is running, but only with an IP that doesn't allow it to communicate. 

I'll post the results later tonight.

Thanks!

**Reply by visualapproach** [reply]
Just make sure you use the dev branch. The master will not work without WiFi after power out. We are discussing some WiFi stuff out in a third branch but I haven't tested that yet so until further you should stick to development.

**Reply by davidmardanielsson** [reply]
Ah ok. Little confused. You mentioned after my last test that I should use master. Should I go back to dev?

**Reply by visualapproach** [reply]
Yes I understand that. Since you were leaving the hut I thought it would be best to use the stable master version instead of the not finished test in the dev. But the "work without WiFi" feat is only in the dev branch. When it's good enough I'll push it to master. So you should try the development version now if your up (out?) there.

**Reply by davidmardanielsson** [reply]
Yep! I'm there for the next few days. Will try dev. Thanks

**Reply by visualapproach** [reply]
NP, sorry to confuse you

**Reply by visualapproach** [reply]
Hi again. The "temp"-branch is now another option to test if you want. It's designed to work without wifi as well, but you just have to edit the file "config.h" and change line 53 to
```const bool enableWmApFallback = false;```
to stop the blocking autoportal from running. You need to have had the ESP connected to WiFi with your credentials before you edit that line and upload. Don't forget to upload the filesystem also.

You don't have to try that branch of course, I just mention it in case you'd like the new web interface by @torei or just want to help out testing.

**Reply by torei** [reply]
@visualapproach sorry to correct you. @davidmardanielsson you just have to upload the filesystem and flash the temp branch. on the first startup, even if you set enableWmApFallback=false, the WiFi manager starts and asks you for your local WiFi credentials. :-)
or you could set your local WiFi credentials in the config as well. then it will try to connect with these instead of starting any AP.

would be nice to have that branch tested, by the way :-)
